### ISSUE TITLE
[LDB-61] star join test hangs

### ISSUE THREAD

**Issue by dynamobi-build** [LucidDB/luciddb#260]
<b><i>[reporter="elin", created="Wed, 12 Apr 2006 11:27:28 -0500 (GMT-05:00)"]</i></b>
Hanging test found during nightly test run, under open/luciddb/test/sql/pge/8.sql<br/>
<br/>
To rerun just this test, mount raksha and then go to open/luciddb/test/sql/pge/<br/>
and run<br/>
<br/>
ant test -Dtestfile=test.xml -Dforeign.db=flatfile -Dbh.test-names=pge.index,pge.8<br/>
<br/>
Log with thread dump:<br/>
0: jdbc:luciddb:> SELECT AL4.REVN_YR, AL4.REVN_MO, AL3.SRVPLN_ID,<br/>
. . . . . . . . >  AL3.SRVPLN_NM, AL2.GEOG_CNTY_CD, SUM(AL5.RSD_SVC_BILG_AMT)<br/>
. . . . . . . . > FROM<br/>
. . . . . . . . >  PGE.CUST_SERV_ACCT AL1,<br/>
. . . . . . . . >  PGE.GEOG AL2,<br/>
. . . . . . . . >  PGE.SERV_PLAN AL3,<br/>
. . . . . . . . >  PGE.REVN_PRD AL4,<br/>
. . . . . . . . >  PGE.REVN_DTL_RAND AL5<br/>
. . . . . . . . >  WHERE (AL5.REVN_YR_MO=AL4.REVN_YR_MO<br/>
. . . . . . . . >  AND AL2.GEOG_KEY=AL5.GEOG_KEY<br/>
. . . . . . . . >  AND AL3.SRVPLN_KEY=AL5.SRVPLN_KEY<br/>
. . . . . . . . >  AND AL1.CUST_SERV_ACCT_KEY=AL5.CUST_SERV_ACCT_KEY)<br/>
. . . . . . . . > GROUP BY AL4.REVN_YR, AL4.REVN_MO, AL3.SRVPLN_ID,<br/>
. . . . . . . . >  AL3.SRVPLN_NM, AL2.GEOG_CNTY_CD<br/>
. . . . . . . . > ORDER BY  5, 1, 2, 3, 4, 6<br/>
. . . . . . . . > ;<br/>
Full thread dump Java HotSpot(TM) Client VM (1.5.0_06-b05 mixed mode, sharing):<br/>
<br/>
"Timer-0" prio=1 tid=0x08595668 nid=0x22ee in Object.wait() [0xa244a000..0xa244af30]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x896f65f0> (a java.util.TaskQueue)<br/>
        at java.util.TimerThread.mainLoop(Timer.java:509)<br/>
        - locked <0x896f65f0> (a java.util.TaskQueue)<br/>
        at java.util.TimerThread.run(Timer.java:462)<br/>
<br/>
"MDR event dispatcher" daemon prio=1 tid=0x08093a68 nid=0x22e7 in Object.wait() [0xb114a000..0xb114afb0]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x8954dd70> (a java.util.LinkedList)<br/>
        at java.lang.Object.wait(Object.java:474)<br/>
        at org.netbeans.mdr.util.EventNotifier$EventsDelivery.run(EventNotifier.java:257)<br/>
        - locked <0x8954dd70> (a java.util.LinkedList)<br/>
        at java.lang.Thread.run(Thread.java:595)<br/>
<br/>
"HSQLDB Timer @1cb25f1" daemon prio=1 tid=0x0849ab30 nid=0x22e4 in Object.wait() [0xb1268000..0xb1268e30]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x89012180> (a org.hsqldb.lib.HsqlTimer)<br/>
        at org.hsqldb.lib.HsqlTimer.nextTask(Unknown Source)<br/>
        - locked <0x89012180> (a org.hsqldb.lib.HsqlTimer)<br/>
        at org.hsqldb.lib.HsqlTimer$TaskRunner.run(Unknown Source)<br/>
        at java.lang.Thread.run(Thread.java:595)<br/>
<br/>
"Active Reference Queue Daemon" daemon prio=1 tid=0x08488b90 nid=0x22e3 in Object.wait() [0xb11e7000..0xb11e7eb0]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x89003cb8> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:116)<br/>
        - locked <0x89003cb8> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at org.openide.util.Utilities$ActiveQueue.run(Utilities.java:2442)<br/>
        at java.lang.Thread.run(Thread.java:595)<br/>
<br/>
"Low Memory Detector" daemon prio=1 tid=0x080a5118 nid=0x22d0 runnable [0x00000000..0x00000000]<br/>
<br/>
"CompilerThread0" daemon prio=1 tid=0x080a3bb8 nid=0x22cf waiting on condition [0x00000000..0xb2287898]<br/>
<br/>
"Signal Dispatcher" daemon prio=1 tid=0x080a2cb0 nid=0x22ce waiting on condition [0x00000000..0x00000000]<br/>
<br/>
"Finalizer" daemon prio=1 tid=0x0809bf70 nid=0x22cd in Object.wait() [0xb24d2000..0xb24d21b0]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x88f85800> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:116)<br/>
        - locked <0x88f85800> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:132)<br/>
        at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:159)<br/>
<br/>
"Reference Handler" daemon prio=1 tid=0x0809b260 nid=0x22cc in Object.wait() [0xb2552000..0xb2553030]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        at java.lang.Object.wait(Object.java:474)<br/>
        at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:116)<br/>
        - locked <0x88f85880> (a java.lang.ref.Reference$Lock)<br/>
<br/>
"main" prio=1 tid=0x0805def0 nid=0x22ca runnable [0xbfdaa000..0xbfdaaac8]<br/>
        at net.sf.farrago.fennel.FennelStorage.tupleStreamFetch(Native Method)<br/>
        at net.sf.farrago.fennel.FennelStreamGraph.fetch(FennelStreamGraph.java:175)<br/>
        at net.sf.farrago.runtime.FennelTupleIter.populateBuffer(FennelTupleIter.java:110)<br/>
        at net.sf.farrago.runtime.FennelAbstractTupleIter.fetchNext(FennelAbstractTupleIter.java:99)<br/>
        at org.eigenbase.runtime.TupleIterResultSet.next(TupleIterResultSet.java:90)<br/>
        at net.sf.farrago.runtime.FarragoTupleIterResultSet.next(FarragoTupleIterResultSet.java:97)<br/>
        at sqlline.SqlLine$BufferedRows.<init>(Unknown Source)<br/>
        at sqlline.SqlLine.print(Unknown Source)<br/>
        at sqlline.SqlLine$Commands.execute(Unknown Source)<br/>
        at sqlline.SqlLine$Commands.sql(Unknown Source)<br/>
        at sqlline.SqlLine.dispatch(Unknown Source)<br/>
        at sqlline.SqlLine.begin(Unknown Source)<br/>
        at sqlline.SqlLine.mainWithInputRedirection(Unknown Source)<br/>
        at sqlline.SqlLine.main(Unknown Source)<br/>
<br/>
"VM Thread" prio=1 tid=0x08098708 nid=0x22cb runnable<br/>
<br/>
"VM Periodic Task Thread" prio=1 tid=0x080a65b8 nid=0x22d1 waiting on condition<br/>
<br/>
Full thread dump Java HotSpot(TM) Client VM (1.5.0_06-b05 mixed mode, sharing):<br/>
<br/>
"Timer-0" prio=1 tid=0x08595668 nid=0x22ee in Object.wait() [0xa244a000..0xa244af30]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x896f65f0> (a java.util.TaskQueue)<br/>
        at java.util.TimerThread.mainLoop(Timer.java:509)<br/>
        - locked <0x896f65f0> (a java.util.TaskQueue)<br/>
        at java.util.TimerThread.run(Timer.java:462)<br/>
<br/>
"MDR event dispatcher" daemon prio=1 tid=0x08093a68 nid=0x22e7 in Object.wait() [0xb114a000..0xb114afb0]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x8954dd70> (a java.util.LinkedList)<br/>
        at java.lang.Object.wait(Object.java:474)<br/>
        at org.netbeans.mdr.util.EventNotifier$EventsDelivery.run(EventNotifier.java:257)<br/>
        - locked <0x8954dd70> (a java.util.LinkedList)<br/>
        at java.lang.Thread.run(Thread.java:595)<br/>
<br/>
"HSQLDB Timer @1cb25f1" daemon prio=1 tid=0x0849ab30 nid=0x22e4 in Object.wait() [0xb1268000..0xb1268e30]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x89012180> (a org.hsqldb.lib.HsqlTimer)<br/>
        at org.hsqldb.lib.HsqlTimer.nextTask(Unknown Source)<br/>
        - locked <0x89012180> (a org.hsqldb.lib.HsqlTimer)<br/>
        at org.hsqldb.lib.HsqlTimer$TaskRunner.run(Unknown Source)<br/>
        at java.lang.Thread.run(Thread.java:595)<br/>
<br/>
"Active Reference Queue Daemon" daemon prio=1 tid=0x08488b90 nid=0x22e3 in Object.wait() [0xb11e7000..0xb11e7eb0]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x89003cb8> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:116)<br/>
        - locked <0x89003cb8> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at org.openide.util.Utilities$ActiveQueue.run(Utilities.java:2442)<br/>
        at java.lang.Thread.run(Thread.java:595)<br/>
<br/>
"Low Memory Detector" daemon prio=1 tid=0x080a5118 nid=0x22d0 runnable [0x00000000..0x00000000]<br/>
<br/>
"CompilerThread0" daemon prio=1 tid=0x080a3bb8 nid=0x22cf waiting on condition [0x00000000..0xb2287898]<br/>
<br/>
"Signal Dispatcher" daemon prio=1 tid=0x080a2cb0 nid=0x22ce runnable [0x00000000..0x00000000]<br/>
<br/>
"Finalizer" daemon prio=1 tid=0x0809bf70 nid=0x22cd in Object.wait() [0xb24d2000..0xb24d21b0]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x88f85800> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:116)<br/>
        - locked <0x88f85800> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:132)<br/>
        at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:159)<br/>
<br/>
"Reference Handler" daemon prio=1 tid=0x0809b260 nid=0x22cc in Object.wait() [0xb2552000..0xb2553030]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        at java.lang.Object.wait(Object.java:474)<br/>
        at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:116)<br/>
        - locked <0x88f85880> (a java.lang.ref.Reference$Lock)<br/>
<br/>
"main" prio=1 tid=0x0805def0 nid=0x22ca runnable [0xbfdaa000..0xbfdaaac8]<br/>
        at net.sf.farrago.fennel.FennelStorage.tupleStreamFetch(Native Method)<br/>
        at net.sf.farrago.fennel.FennelStreamGraph.fetch(FennelStreamGraph.java:175)<br/>
        at net.sf.farrago.runtime.FennelTupleIter.populateBuffer(FennelTupleIter.java:110)<br/>
        at net.sf.farrago.runtime.FennelAbstractTupleIter.fetchNext(FennelAbstractTupleIter.java:99)<br/>
        at org.eigenbase.runtime.TupleIterResultSet.next(TupleIterResultSet.java:90)<br/>
        at net.sf.farrago.runtime.FarragoTupleIterResultSet.next(FarragoTupleIterResultSet.java:97)<br/>
        at sqlline.SqlLine$BufferedRows.<init>(Unknown Source)<br/>
        at sqlline.SqlLine.print(Unknown Source)<br/>
        at sqlline.SqlLine$Commands.execute(Unknown Source)<br/>
        at sqlline.SqlLine$Commands.sql(Unknown Source)<br/>
        at sqlline.SqlLine.dispatch(Unknown Source)<br/>
        at sqlline.SqlLine.begin(Unknown Source)<br/>
        at sqlline.SqlLine.mainWithInputRedirection(Unknown Source)<br/>
        at sqlline.SqlLine.main(Unknown Source)<br/>
<br/>
"VM Thread" prio=1 tid=0x08098708 nid=0x22cb runnable<br/>
<br/>
"VM Periodic Task Thread" prio=1 tid=0x080a65b8 nid=0x22d1 waiting on condition<br/>
<br/>
Full thread dump Java HotSpot(TM) Client VM (1.5.0_06-b05 mixed mode, sharing):<br/>
<br/>
"Timer-0" prio=1 tid=0x08595668 nid=0x22ee in Object.wait() [0xa244a000..0xa244af30]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x896f65f0> (a java.util.TaskQueue)<br/>
        at java.util.TimerThread.mainLoop(Timer.java:509)<br/>
        - locked <0x896f65f0> (a java.util.TaskQueue)<br/>
        at java.util.TimerThread.run(Timer.java:462)<br/>
<br/>
"MDR event dispatcher" daemon prio=1 tid=0x08093a68 nid=0x22e7 in Object.wait() [0xb114a000..0xb114afb0]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x8954dd70> (a java.util.LinkedList)<br/>
        at java.lang.Object.wait(Object.java:474)<br/>
        at org.netbeans.mdr.util.EventNotifier$EventsDelivery.run(EventNotifier.java:257)<br/>
        - locked <0x8954dd70> (a java.util.LinkedList)<br/>
        at java.lang.Thread.run(Thread.java:595)<br/>
<br/>
"HSQLDB Timer @1cb25f1" daemon prio=1 tid=0x0849ab30 nid=0x22e4 in Object.wait() [0xb1268000..0xb1268e30]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x89012180> (a org.hsqldb.lib.HsqlTimer)<br/>
        at org.hsqldb.lib.HsqlTimer.nextTask(Unknown Source)<br/>
        - locked <0x89012180> (a org.hsqldb.lib.HsqlTimer)<br/>
        at org.hsqldb.lib.HsqlTimer$TaskRunner.run(Unknown Source)<br/>
        at java.lang.Thread.run(Thread.java:595)<br/>
<br/>
"Active Reference Queue Daemon" daemon prio=1 tid=0x08488b90 nid=0x22e3 in Object.wait() [0xb11e7000..0xb11e7eb0]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x89003cb8> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:116)<br/>
        - locked <0x89003cb8> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at org.openide.util.Utilities$ActiveQueue.run(Utilities.java:2442)<br/>
        at java.lang.Thread.run(Thread.java:595)<br/>
<br/>
"Low Memory Detector" daemon prio=1 tid=0x080a5118 nid=0x22d0 runnable [0x00000000..0x00000000]<br/>
<br/>
"CompilerThread0" daemon prio=1 tid=0x080a3bb8 nid=0x22cf waiting on condition [0x00000000..0xb2287898]<br/>
<br/>
"Signal Dispatcher" daemon prio=1 tid=0x080a2cb0 nid=0x22ce runnable [0x00000000..0x00000000]<br/>
<br/>
"Finalizer" daemon prio=1 tid=0x0809bf70 nid=0x22cd in Object.wait() [0xb24d2000..0xb24d21b0]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        - waiting on <0x88f85800> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:116)<br/>
        - locked <0x88f85800> (a java.lang.ref.ReferenceQueue$Lock)<br/>
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:132)<br/>
        at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:159)<br/>
<br/>
"Reference Handler" daemon prio=1 tid=0x0809b260 nid=0x22cc in Object.wait() [0xb2552000..0xb2553030]<br/>
        at java.lang.Object.wait(Native Method)<br/>
        at java.lang.Object.wait(Object.java:474)<br/>
        at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:116)<br/>
        - locked <0x88f85880> (a java.lang.ref.Reference$Lock)<br/>
<br/>
"main" prio=1 tid=0x0805def0 nid=0x22ca runnable [0xbfdaa000..0xbfdaaac8]<br/>
        at net.sf.farrago.fennel.FennelStorage.tupleStreamFetch(Native Method)<br/>
        at net.sf.farrago.fennel.FennelStreamGraph.fetch(FennelStreamGraph.java:175)<br/>
        at net.sf.farrago.runtime.FennelTupleIter.populateBuffer(FennelTupleIter.java:110)<br/>
        at net.sf.farrago.runtime.FennelAbstractTupleIter.fetchNext(FennelAbstractTupleIter.java:99)<br/>
        at org.eigenbase.runtime.TupleIterResultSet.next(TupleIterResultSet.java:90)<br/>
        at net.sf.farrago.runtime.FarragoTupleIterResultSet.next(FarragoTupleIterResultSet.java:97)<br/>
        at sqlline.SqlLine$BufferedRows.<init>(Unknown Source)<br/>
        at sqlline.SqlLine.print(Unknown Source)<br/>
        at sqlline.SqlLine$Commands.execute(Unknown Source)<br/>
        at sqlline.SqlLine$Commands.sql(Unknown Source)<br/>
        at sqlline.SqlLine.dispatch(Unknown Source)<br/>
        at sqlline.SqlLine.begin(Unknown Source)<br/>
        at sqlline.SqlLine.mainWithInputRedirection(Unknown Source)<br/>
        at sqlline.SqlLine.main(Unknown Source)<br/>
<br/>
"VM Thread" prio=1 tid=0x08098708 nid=0x22cb runnable<br/>
<br/>
"VM Periodic Task Thread" prio=1 tid=0x080a65b8 nid=0x22d1 waiting on condition<br/>
<br/>
<br/>
<br/>
<br/>

**Comment by dynamobi-build** [comment]
<b><i>[author="zfong", created="Wed, 12 Apr 2006 13:21:06 -0500 (GMT-05:00)"]</i></b>
The query appears to be running a long time because no semijoins are being chosen to execute the query.  I have not had a chance to determine if that's a limitation in the current costing used by the optimizer or a legitimate bug in the optimizer not selecting a semijoin when it should be.

**Comment by dynamobi-build** [comment]
<b><i>[author="boris", created="Wed, 12 Apr 2006 18:16:58 -0500 (GMT-05:00)"]</i></b>
<br/>
I think this might not be a hang. I ran it, and it was running for about 1.5 hours. And seems like cpu and mem is active. Liz mentioned the datasize is large, so it's probably to be expected. <br/>
<br/>
I'll downgrade this for now, and we can look at it when the hashjoins are in.<br/>
<br/>
We're also going to add a new test attribute to allow us to filter tests for features not quite implemented yet, such as this.<br/>
<br/>

**Comment by dynamobi-build** [comment]
<b><i>[author="zfong", created="Mon, 17 Apr 2006 15:29:53 -0500 (GMT-05:00)"]</i></b>
My previous comment indicating that semijoins weren't being used was incorrect.  There must have been some problem in my environment, as when I reran the query today, I am now seeing semijoins.  However, given the way the query is written, the wrong semijoins are being chosen.  The fact table in the query is "REVN_DTL_RAND".  Since we don't have real stats yet, in order to force the optimizer to pick this table as the fact table, it needs to appear FIRST in the from clause.  <br/>
<br/>
When I changed the query to move that table to the first element in the FROM clause, I ran into another bug.  The fix for that bug is in Eigenchange 6258.  With that fix, REVN_DTL_RAND is now chosen as the fact table, with the other four tables serving as dimension tables filtering that table.<br/>
<br/>
However, because there are no table level filters on the dimension table, the result of the join is still fairly large, and therefore, without hash joins, the query will do a large cartesian product join, and as a result, will take a while to run.
